neuralpredictors/training/tracking.py

```diff
@@ -146,3 +154,369 @@ def asdict(self, time_key="time", make_copy=True):
+            if isinstance(objective, abc.Mapping):
+                self._initialize_epoch(log[key], objective)
+            elif not callable(objective):
+                while len(log[key]) <= self.epoch:
+                    log[key].append(objective)
+
+    def _log_objective_value(self, value: float, log: Union[Mapping, Sequence], key: Tuple[str, ...] = ()) -> None:
+        """
+        Recursively walk through the log dictionary to get to follow `key`.
+        When on lowest level: add `value` to entry at current epoch.
+        Args:
+            value: objective value to log
+            log: log subdict to add value to
+            key: key for where `value` is saved
+        """
+        if len(key) > 1:
+            self._log_objective_value(value, log[key[0]], key[1:])
+        else:
+            log[key[0]][self.epoch] += value
+
+    def _log_objective_callables(self, log: Union[Mapping, Sequence], objectives: Mapping[str, any]) -> None:
+        """
+        Log all objectives that are specified as callables
+        Disclaimer: this is not very well tested!
+        Args:
+            log:
+            objectives:
+        """
+        for key, objective in objectives.items():
+            if isinstance(objective, abc.Mapping):
+                self._log_objective_callables(log[key], objective)
+            elif callable(objective):
+                log[key][self.epoch] += objective()
+
+    def _normalize_log(self, log: Union[Mapping, Sequence]) -> Union[Mapping, np.array]:
+        """
+        Recursively go through the log and normalize the entries that
+        have `"normalization"` information on the same level.
+        Args:
+            log: subdict on which to apply normalization or list (if on lowest level)
+
+        Returns:
+            Normalized log dictionary that has numpy arrays on the lowest level
+
+        """
+        if isinstance(log, abc.Mapping):
+            n_log = {}
+            norm = None
+            for key, l in log.items():
+                res = self._normalize_log(l)  # to turn into arrays
+                if key == "normalization":
+                    assert isinstance(res, np.ndarray)
+                    norm = res
+                else:
+                    n_log[key] = res
+            self._normalize(n_log, norm)
+            return n_log
+        else:
+            return np.array(log)
+
+    def _normalize(self, n_log: Mapping, norm: np.ndarray):
+        """
+        Normalizes sub-dict `n_log`
+        """
+        if norm is not None:
+            nonzero_start = (norm != 0).argmax(axis=0)
+            norm = norm[nonzero_start:]
+            for key, l in n_log.items():
+                l = l[nonzero_start:]
+                if isinstance(l, np.ndarray):
+                    n_log[key] = l / np.where(norm > 0, norm, np.ones_like(norm))
+
+    def _gather_log(self, log: Union[Mapping, Sequence], key: Tuple[str, ...], index: int = -1) -> Mapping[str, str]:
```

Comment:
I would say `Dict[str, str]` as the type of the return value because the method always returns a dictionary. That way the caller doesn't have to check whether it is a `Dict` or not (if that should matter to the caller).